larq_zoo/sota/quicknet.py

```diff
@@ -104,6 +161,21 @@ def build(self) -> tf.keras.models.Model:
 
         model = tf.keras.Model(inputs=self.image_input, outputs=x, name="quicknet")
 
+        return model
+
+
+@factory
+class QuickNetFactory(QuickNetBaseFactory):
+    """Quicknet - A model designed for fast inference using [Larq Compute Engine](https://github.com/larq/compute-engine)"""
+
+    spec = Field(
+        lambda _: ([2, 3, 4, 4], [64, 128, 256, 512], [False, False, False, False])
```

Comment:
Not sure why the `lambda` is necessary here. @AdamHillier, is this the intended usage or do you have a suggestion?

Reply:
It's necessary because anything that's immutable has to be wrapped in a lambda, or *bad things can happen*™. (Bad things in this case means subtle bugs arising from mutable data being modified in one instance of the component class with unpredictable results in other instances. If we require a lambda which returns the data, we can safely generate a new value of the data in each class instance.)

This is a tuple of lists, and lists are mutable, hence the lambda is required. If it were a tuple of tuples of immutable values it would technically be fine, but Zookeeper only checks one level of recursion down and otherwise conservatively assumes mutability, requiring a lambda. The gist is that Zookeeper knows that `None`, `bool`, `int`, `float`, and `str` are all immutable. It also knows that `tuple`s and `frozenset`s of these fundamental types are immutable, and assumes that everything else is mutable.

The lambda can accept 0 or 1 argument (`self`) by the way, so there's no need for the underscore here.

Reply:
Yes this is intended behaviour.
`Field` doesn't allow mutable default values since this could lead to unexpected behaviour. This is inline with what Python dataclasses do. Take a look at https://docs.python.org/3/library/dataclasses.html#mutable-default-values for an example why mutable default values would be a bad idea.

You could get rid of the lambda here if you use a [`frozenset`](https://docs.python.org/3/library/stdtypes.html#frozenset)  or a tuple of `frozenset`s. Though the lambda is also OK.

Reply:
Okay thanks! I had tried a tuple of tuples, this explains why that throws as well.